### ISSUE TITLE
MNN initialization with large number of batches fails

### ISSUE THREAD

**Issue by stanleyjs** [KrishnaswamyLab/graphtools#12]

**Comment by stanleyjs** [comment]
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in K(self)
>     334         try:
> --> 335             return self._kernel
>     336         except AttributeError:
> 
> AttributeError: 'Graph' object has no attribute '_kernel'
> 
> During handling of the above exception, another exception occurred:
> 
> IndexError                                Traceback (most recent call last)
> <ipython-input-125-deadd7696fda> in <module>()
> ----> 1 g_mnn = gt.Graph(combined[newidx,:],use_pygsp=True, knn=5, sample_idx = clusts,gamma = 0.1)
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/api.py in Graph(data, n_pca, sample_idx, adaptive_k, precomputed, knn, decay, distance, thresh, n_landmark, n_svd, beta, gamma, n_jobs, verbose, random_state, graphtype, use_pygsp, initialize, **kwargs)
>     246                  random_state=random_state,
>     247                  initialize=initialize,
> --> 248                  **kwargs)
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/graphs.py in __init__(self, data, sample_idx, knn, beta, gamma, n_pca, adaptive_k, **kwargs)
>     847                 raise ValueError("gamma must be a symmetric matrix")
>     848 
> --> 849         super().__init__(data, n_pca=n_pca, **kwargs)
>     850 
>     851     def _weight_knn(self, sample_size=None):
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in __init__(self, data, verbose, n_jobs, **kwargs)
>     457         self.verbose = verbose
>     458         set_logging(verbose)
> --> 459         super().__init__(data, **kwargs)
>     460 
>     461     def get_params(self):
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in __init__(self, data, n_pca, random_state, **kwargs)
>      85         self.random_state = random_state
>      86         self.data_nu = self._reduce_data()
> ---> 87         super().__init__(**kwargs)
>      88 
>      89     def _reduce_data(self):
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in __init__(self, initialize, **kwargs)
>     249         if initialize:
>     250             log_debug("Initializing kernel...")
> --> 251             self.K
>     252         else:
>     253             log_debug("Not initializing kernel.")
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in K(self)
>     335             return self._kernel
>     336         except AttributeError:
> --> 337             self._kernel = self._build_kernel()
>     338             return self._kernel
>     339 
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/base.py in _build_kernel(self)
>     268         RuntimeWarning : if K is not symmetric
>     269         """
> --> 270         kernel = self.build_kernel()
>     271         if (kernel - kernel.T).max() > 1e-5:
>     272             warnings.warn("K should be symmetric", RuntimeWarning)
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/graphs.py in build_kernel(self)
>     990                 Kij = Y.build_kernel_to_data(
>     991                     X.data_nu,
> --> 992                     knn=self.weighted_knn[i])
>     993                 if i == j:
>     994                     # downweight within-batch affinities by beta
> 
> ~/anaconda3/lib/python3.6/site-packages/graphtools-0.1.3-py3.6.egg/graphtools/graphs.py in build_kernel_to_data(self, Y, knn)
>     222             log_complete("KNN search")
>     223             log_start("affinities")
> --> 224             bandwidth = distances[:, knn - 1]
>     225             radius = bandwidth * np.power(-1 * np.log(self.thresh),
>     226                                           1 / self.decay)
> 
> IndexError: index 10 is out of bounds for axis 1 with size 7
>

**Comment by stanleyjs** [comment]
the offending function calls are in the "sparse fast alpha decay" section of `build_kernel_to_data` @scottgigante

**Comment by stanleyjs** [comment]
the batch distribution that produced this error was 
`array([ 31., 197.,  93., 132.,  83.,  97.,  33.,  75.,  58.,  61.,  18.,
         75.,  26.,  21.])`

**Comment by stanleyjs** [comment]
a similar situation can be spawned via knn:
`ValueError: Expected n_neighbors <= n_samples,  but n_samples = 7, n_neighbors = 11`

**Comment by stanleyjs** [comment]
TL;DR: This issue arises from `_weight_knn`.  The default parameter  `adaptive_k = 'sqrt'` will make this occur.  A workaround is `adaptive_k = 'none'`

**Comment by scottgigante** [comment]
Can you try again with `verbose=2` and post the logging output?

**Comment by scottgigante** [comment]
`adaptive_k` has been deprecated. Closing this issue.
